Tests/CombineWampTests/IntegrationTests/SubscriberTests.swift

```diff
@@ -0,0 +1,115 @@
+// https://crossbar.io/docs/Getting-Started/#publishing-client
+// All of this will eventually be automated.
+final class SubscriberTests: IntegrationTestBase {
+    var connected: XCTestExpectation!
+    var session: WampSession!
+    let realm = URI("realm1")!
+    var connection: AnyCancellable?
+
+    override func setUp() {
+        super.setUp()
+
+        connected = expectation(description: "Connected")
+        session = WampSession(transport: transport(), serialization: serialization, realm: realm, roles: .allClientRoles)
+        connection = session.connect()
+            .sink(
+                receiveCompletion: { completion in
+                    switch completion {
+                    case let .failure(error):
+                        XCTFail(error.localizedDescription)
+                    case .finished:
+                        break
+                    }
+                },
+                receiveValue: { [weak self] welcome in
+                    self?.connected.fulfill()
+                }
+            )
+    }
+
+    func testSubscribeAndReceiveEvents() throws {
+        wait(for: [connected], timeout: 0.5)
+
+        let receiveEventsExpectation = expectation(description: "3 events should have been received")
+        receiveEventsExpectation.expectedFulfillmentCount = 3
+        var receivedEvents = [Message.Event]()
+
+        session
+            .client
+            .asSubscriber?
+            .subscribe(topic: URI("com.myapp.hello")!, onUnsubscribe: { _ in })
+            .sink(
+                receiveCompletion: { completion in
+                    switch completion {
+                    case let .failure(error):
+                        XCTFail(error.localizedDescription)
+                    case .finished:
+                        break
+                    }
+                },
+                receiveValue: { event in
+                    receivedEvents.append(event)
+                    receiveEventsExpectation.fulfill()
+                }
+            ).store(in: &cancellables)
+
+        wait(for: [receiveEventsExpectation], timeout: 5.0)
+        XCTAssertEqual(receivedEvents.count, 3)
+        for i in 0..<3 {
+            XCTAssertNil(receivedEvents[i].argumentsKw)
+            XCTAssertEqual(receivedEvents[i].arguments?.count, 1)
+            XCTAssertTrue(receivedEvents[i].arguments?.first?.string?.hasPrefix("Hello World ") == true)
+        }
+        XCTAssertNotNil(connection)
+    }
+
+    func testSubscribeAndUnsubscribe() throws {
+        wait(for: [connected], timeout: 0.5)
+
+        let receiveEventsExpectation = expectation(description: "Events should have been received")
+        let unsubscribingExpectation = expectation(description: "Unsubscribed ack should have been received")
+        receiveEventsExpectation.assertForOverFulfill = false
+
+        let subscription = session
+            .client
+            .asSubscriber?
```

Comment:
same thinking here

Reply:
Yeah, and probably for all the other stuff... I'm gonna change all these things in an upcoming PR, it's easier for me.